File: lib/biotp-py/biotp/ecdh_totp.py
"""
ECDH-TOTP: Elliptic Curve Diffie-Hellman Time-Based One-Time Passwords.

Implements the key derivation, shared secret computation, and OTP
generation/verification described in rfc-ecdh-totp.txt.
"""

import hashlib
import hmac
import struct
import time

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.kdf.x963kdf import X963KDF
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

P256_ORDER = (
    0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551
)


class MasterKey:
    """Server-side ECDH-TOTP master key.

    Holds a P-256 master private scalar derived from a seed via HKDF and
    exposes child-key derivation, shared-secret computation, and OTP
    generation/verification.
    """

    def __init__(self, secret: bytes, *, period: int = 30) -> None:
        self._secret = secret
        self._period = period

        raw = self._derive_material(b"humancheck:master-private")
        self._scalar = int.from_bytes(raw, "big") % (P256_ORDER - 1) + 1
        self._private_key = ec.derive_private_key(self._scalar, ec.SECP256R1())

        uncompressed = self._private_key.public_key().public_bytes(
            Encoding.X962, PublicFormat.UncompressedPoint
        )
        self._public_raw = uncompressed[1:]  # raw X||Y (64 bytes)

    # -- properties ----------------------------------------------------------

    @property
    def public_raw(self) -> bytes:
        """Master public key as raw X||Y (64 bytes)."""
        return self._public_raw

    @property
    def period(self) -> int:
        return self._period

    # -- time ----------------------------------------------------------------

    def current_counter(self) -> int:
        return int(time.time()) // self._period

    # -- child key derivation ------------------------------------------------

    def tweak_scalar(self, counter: int) -> int:
        counter_bytes = struct.pack(">Q", counter)
        digest = hmac.new(
            self._public_raw, counter_bytes, hashlib.sha256
        ).digest()
        return int.from_bytes(digest, "big") % P256_ORDER

    def child_private_for_counter(
        self, counter: int
    ) -> ec.EllipticCurvePrivateKey:
        d = (self._scalar + self.tweak_scalar(counter)) % P256_ORDER
        if d == 0:
            d = 1
        return ec.derive_private_key(d, ec.SECP256R1())

    def child_pubkey_hex(self, counter: int) -> str:
        return (
            self.child_private_for_counter(counter)
            .public_key()
            .public_bytes(Encoding.X962, PublicFormat.UncompressedPoint)
            .hex()
        )

    # -- shared secret & OTP -------------------------------------------------

    @staticmethod
    def compute_shared_secret(
        server_private: ec.EllipticCurvePrivateKey,
        user_pubkey_hex: str,
    ) -> bytes:
        user_pub = ec.EllipticCurvePublicKey.from_encoded_point(
            ec.SECP256R1(), bytes.fromhex(user_pubkey_hex)
        )
        raw_shared = server_private.exchange(ec.ECDH(), user_pub)
        return X963KDF(
            algorithm=hashes.SHA256(),
            length=32,
            sharedinfo=b"",
        ).derive(raw_shared)

    @staticmethod
    def compute_totp(shared_secret: bytes, counter: int) -> str:
        counter_bytes = struct.pack(">Q", counter)
        h = hmac.new(shared_secret, counter_bytes, hashlib.sha256).digest()
        offset = h[-1] & 0x0F
        truncated = (
            struct.unpack(">I", h[offset : offset + 4])[0] & 0x7FFFFFFF
        )
        otp = truncated % 1_000_000
        return f"{otp:06d}"

    def verify_otp(
        self,
        user_pubkey_hex: str,
        otp: str,
        *,
        skew: int = 1,
        last_counter: int = -1,
    ) -> tuple[bool, int]:
        """Verify an OTP against the current time window.

        Returns (valid, matched_counter).  If invalid, matched_counter is -1.
        """
        counter = self.current_counter()
        candidates = [counter + i for i in range(-skew, skew + 1)]
        for c in candidates:
            if c <= last_counter:
                continue
            sk = self.child_private_for_counter(c)
            shared = self.compute_shared_secret(sk, user_pubkey_hex)
            expected = self.compute_totp(shared, c)
            if hmac.compare_digest(expected, otp):
                return True, c
        return False, -1

    # -- internal ------------------------------------------------------------

    def _derive_material(self, label: bytes, length: int = 32) -> bytes:
        return HKDF(
            algorithm=hashes.SHA256(),
            length=length,
            salt=b"humancheck-offline-v1",
            info=label,
        ).derive(self._secret)

File: lib/biotp-ts/src/ecdh_totp.ts
import * as crypto from "node:crypto";

const P256_ORDER = BigInt(
  "0xFFFFFFFF00000000FFFFFFFFFFFFFFFFBCE6FAADA7179E84F3B9CAC2FC632551"
);

function bigintToBuffer(n: bigint, length: number): Buffer {
  const hex = n.toString(16).padStart(length * 2, "0");
  return Buffer.from(hex, "hex");
}

function bufferToBigint(buf: Buffer): bigint {
  return BigInt("0x" + buf.toString("hex"));
}

function x963kdf(
  sharedSecret: Buffer,
  sharedInfo: Buffer,
  length: number
): Buffer {
  const blocks: Buffer[] = [];
  let total = 0;
  let counter = 1;
  while (total < length) {
    const counterBuf = Buffer.alloc(4);
    counterBuf.writeUInt32BE(counter);
    const hash = crypto
      .createHash("sha256")
      .update(sharedSecret)
      .update(counterBuf)
      .update(sharedInfo)
      .digest();
    blocks.push(hash);
    total += hash.length;
    counter++;
  }
  return Buffer.concat(blocks).subarray(0, length);
}

function ecKeyFromScalar(scalar: bigint): {
  privateKey: crypto.KeyObject;
  publicRaw: Buffer;
} {
  const dBuf = bigintToBuffer(scalar, 32);
  const ecdh = crypto.createECDH("prime256v1");
  ecdh.setPrivateKey(dBuf);
  const pubUncompressed = Buffer.from(ecdh.getPublicKey()); // 04 || X || Y

  const x = pubUncompressed.subarray(1, 33);
  const y = pubUncompressed.subarray(33, 65);

  const privateKey = crypto.createPrivateKey({
    key: {
      kty: "EC",
      crv: "P-256",
      d: dBuf.toString("base64url"),
      x: x.toString("base64url"),
      y: y.toString("base64url"),
    },
    format: "jwk",
  });

  return { privateKey, publicRaw: pubUncompressed };
}

export class MasterKey {
  private _secret: Buffer;
  private _period: number;
  private _scalar: bigint;
  private _privateKey: crypto.KeyObject;
  private _publicRaw: Buffer; // raw X||Y (64 bytes, no 04 prefix)

  constructor(secret: Buffer, period = 30) {
    this._secret = secret;
    this._period = period;

    const raw = this._deriveMaterial(Buffer.from("humancheck:master-private"));
    this._scalar = (bufferToBigint(raw) % (P256_ORDER - 1n)) + 1n;

    const { privateKey, publicRaw } = ecKeyFromScalar(this._scalar);
    this._privateKey = privateKey;
    this._publicRaw = publicRaw.subarray(1); // strip 04 prefix -> X||Y
  }

  get publicRaw(): Buffer {
    return this._publicRaw;
  }

  get period(): number {
    return this._period;
  }

  currentCounter(): number {
    return Math.floor(Date.now() / 1000 / this._period);
  }

  tweakScalar(counter: number): bigint {
    const counterBuf = Buffer.alloc(8);
    counterBuf.writeBigUInt64BE(BigInt(counter));
    const digest = crypto
      .createHmac("sha256", this._publicRaw)
      .update(counterBuf)
      .digest();
    return bufferToBigint(digest) % P256_ORDER;
  }

  childPrivateForCounter(counter: number): crypto.KeyObject {
    let d = (this._scalar + this.tweakScalar(counter)) % P256_ORDER;
    if (d === 0n) d = 1n;
    return ecKeyFromScalar(d).privateKey;
  }

  childPubkeyHex(counter: number): string {
    let d = (this._scalar + this.tweakScalar(counter)) % P256_ORDER;
    if (d === 0n) d = 1n;
    const { publicRaw } = ecKeyFromScalar(d);
    return publicRaw.toString("hex"); // includes 04 prefix
  }

  static computeSharedSecret(
    serverPrivate: crypto.KeyObject,
    userPubkeyHex: string
  ): Buffer {
    const pubBytes = Buffer.from(userPubkeyHex, "hex");
    let x: Buffer, y: Buffer;
    if (pubBytes.length === 65 && pubBytes[0] === 0x04) {
      x = pubBytes.subarray(1, 33);
      y = pubBytes.subarray(33, 65);
    } else if (pubBytes.length === 64) {
      x = pubBytes.subarray(0, 32);
      y = pubBytes.subarray(32, 64);
    } else {
      throw new Error("invalid public key length");
    }

    const userPub = crypto.createPublicKey({
      key: {
        kty: "EC",
        crv: "P-256",
        x: x.toString("base64url"),
        y: y.toString("base64url"),
      },
      format: "jwk",
    });

    const rawShared = crypto.diffieHellman({
      privateKey: serverPrivate,
      publicKey: userPub,
    });

    return x963kdf(Buffer.from(rawShared), Buffer.alloc(0), 32);
  }

  static computeTotp(sharedSecret: Buffer, counter: number): string {
    const counterBuf = Buffer.alloc(8);
    counterBuf.writeBigUInt64BE(BigInt(counter));
    const h = crypto
      .createHmac("sha256", sharedSecret)
      .update(counterBuf)
      .digest();
    const offset = h[h.length - 1] & 0x0f;
    const truncated = h.readUInt32BE(offset) & 0x7fffffff;
    const otp = truncated % 1_000_000;
    return otp.toString().padStart(6, "0");
  }

  verifyOtp(
    userPubkeyHex: string,
    otp: string,
    { skew = 1, lastCounter = -1 }: { skew?: number; lastCounter?: number } = {}
  ): [boolean, number] {
    const counter = this.currentCounter();
    for (let i = -skew; i <= skew; i++) {
      const c = counter + i;
      if (c <= lastCounter) continue;
      const sk = this.childPrivateForCounter(c);
      const shared = MasterKey.computeSharedSecret(sk, userPubkeyHex);
      const expected = MasterKey.computeTotp(shared, c);
      if (
        expected.length === otp.length &&
        crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(otp))
      ) {
        return [true, c];
      }
    }
    return [false, -1];
  }

  private _deriveMaterial(label: Buffer, length = 32): Buffer {
    return Buffer.from(
      crypto.hkdfSync(
        "sha256",
        this._secret,
        Buffer.from("humancheck-offline-v1"),
        label,
        length
      )
    );
  }
}
